### PR TITLE
Dobe/projectionvisitor

### PR DESCRIPTION
**Pull request by dobe**

### CODE REVIEW

File: sql/src/main/java/io/crate/planner/projection/builder/InputCreatingVisitor.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+
+
+        Context(Collection<? extends Symbol> inputs, @Nullable Aggregation.Step fromStep) {
+            this.inputs = new HashMap(inputs.size());
+            int i = 0;
+            for (Symbol input : inputs) {
+                DataType valueType = input.valueType();
+                if (fromStep != null && fromStep == Aggregation.Step.PARTIAL){
+                    // TODO: once we have the datatype of partial aggs, we need to add it here
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/planner/projection/builder/InputCreatingVisitor.java]: we already have that - see AverageAggregation for example

File: sql/src/main/java/io/crate/planner/projection/builder/LeafVisitor.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    public static final LeafVisitor INSTANCE = new LeafVisitor();
+
+    static class Context {
+        final ArrayList<Symbol> leaves;
+
+        Context(ArrayList<Symbol> leaves) {
+            this.leaves = leaves;
+        }
+
+        void allocateLeafeSymbol(Symbol symbol){
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/planner/projection/builder/LeafVisitor.java]: typo

**Comment by mfelsche** [comment]
otherwise ack

**Comment by lmillar** [comment]
I would like to get 30mg adderall and 40mg merhadone

**Pull request merged**